packages/workflow/src/Workflow.ts

```diff
@@ -1421,6 +1424,74 @@ export class Workflow {
 			};
 		}
 	}
+
+	/**
+	 * Select the node to use as starter for a subworkflow execution.
+	 */
+	static selectSubworkflowStarter(nodes: INode[]) {
+		const found = nodes.find((n) => SUBWORKFLOW_STARTER_NODES.includes(n.type));
+
+		if (found) return found;
+
+		const starterNodesToDisplay = SUBWORKFLOW_STARTER_NODES.join(', ');
+
+		throw new ApplicationError(
+			`Missing node to start execution. Please ensure the workflow you are calling contains at least one of the following nodes: ${starterNodesToDisplay}`,
+		);
+	}
+
+	/**
+	 * Select the pinned activator node to use as starter for a manual execution.
+	 *
+	 * In a full manual execution, select the pinned activator that was first added
+	 * to the workflow, prioritizing `n8n-nodes-base.webhook` over other activators.
+	 *
+	 * In a partial manual execution, if the executed node has parent nodes among the
+	 * pinned activators, select the pinned activator that was first added to the workflow,
+	 * prioritizing `n8n-nodes-base.webhook` over other activators. If the executed node
+	 * has no upstream nodes and is itself is a pinned activator, select it.
+	 */
+	selectPinnedActivatorStarter(startNodeNames: string[]) {
+		const allPinnedActivators = this.findAllPinnedActivators();
+
+		const [firstPinnedActivator] = allPinnedActivators;
+
+		// full manual execution
+
+		if (startNodeNames?.length === 0) return firstPinnedActivator ?? null;
+
+		// partial manual execution
+
+		const [firstStartNodeName] = startNodeNames;
+
+		const parentNodeNames = this.getParentNodes(firstStartNodeName);
+
+		if (parentNodeNames.length > 0) {
+			const parentNodeName = parentNodeNames.find((p) => p === firstPinnedActivator.name);
+
+			return allPinnedActivators.find((pa) => pa.name === parentNodeName) ?? null;
+		}
+
+		return allPinnedActivators.find((pa) => pa.name === firstStartNodeName) ?? null;
+	}
+
+	/**
+	 * Find all pinned nodes that may activate a workflow, prioritizing `n8n-nodes-base.webhook`.
+	 */
+	private findAllPinnedActivators() {
+		return this.nodeNames
+			.reduce<INode[]>((acc, nodeName) => {
+				acc.push(this.nodes[nodeName]);
+				return acc;
+			}, [])
+			.filter(
+				(node) =>
+					!node.disabled &&
+					this.pinData?.[node.name] &&
+					['trigger', 'webhook'].some((suffix) => node.type.toLowerCase().endsWith(suffix)),
```

Comment:
```suggestion
					['trigger', 'webhook'].some((suffix) => node.type.toLowerCase().includes(suffix)),
```

I saw this change, was it deliberate? Also one thing I realized is that the `Respond To Webhook` node would fall into this category as eligible trigger node. Ideally we'd have to check the `nodeTypes` whether the node implements the `webhook` or `trigger` functions instead of simply checking by name.

Reply:
>I saw this change, was it deliberate?

Returning `true` when finding `suffix` in non-end position is an error. We have [rules](https://github.com/ivov/eslint-plugin-n8n-nodes-base/blob/master/docs/rules/node-class-description-display-name-unsuffixed-trigger-node.md) that enforce this.

>Also one thing I realized is that the `Respond To Webhook` node would fall into this category as eligible trigger node.

`Respond To Webhook` will return `true` for the check.

>Ideally we'd have to check the `nodeTypes` whether the node implements the `webhook` or `trigger` functions instead of simply checking by name.

Let me take a look. Only moved preexisting logic here.

Reply:
I think we'll have to add an exception since `Respond to Webhook` is not an eligible trigger but would be returned by this, effectively being wrong.

About the usage of `group` you're correct, it's not enforced. Maybe `endsWith && is not respond to webhook`? I know it doesn't look nice but it would at least be correct.